File: yiyue31-translator/scripts/doc_segmenter/reporter.py
"""ReportGenerator implementation.

Generates report.md from SplitContext with source info, split results,
operations, validation results, and file list.
"""

from __future__ import annotations

import os

from doc_segmenter.models import SplitContext


class ReportGeneratorImpl:
    """Implementation of ReportGenerator protocol."""

    def generate_report(self, context: SplitContext) -> None:
        """Generate report.md from the full pipeline context.

        Args:
            context: The complete SplitContext with all pipeline data.
        """
        output_dir = context.output_dir
        os.makedirs(output_dir, exist_ok=True)

        lines = []
        lines.append("# File Split Report")
        lines.append("")

        # Source file information
        lines.append("## Source Information")
        lines.append("")
        lines.append("| Property | Value |")
        lines.append("|----------|-------|")
        lines.append("| File | {} |".format(context.source_info.file_path))
        lines.append("| Size | {:.1f} KB |".format(context.source_info.file_size))
        lines.append("| Lines | {} |".format(context.source_info.file_lines))
        lines.append("| Chars | {} |".format(context.source_info.file_chars))
        lines.append("| Encoding | {} |".format(context.source_info.file_encoding))
        lines.append("")

        # Split results
        lines.append("## Split Results")
        lines.append("")
        lines.append("Total chunks: {}".format(len(context.chunks)))
        lines.append("")

        if context.chunks:
            lines.append("| # | Section | Size (KB) | Lines | Merged |")
            lines.append("|---|---------|-----------|-------|--------|")
            for i, chunk in enumerate(context.chunks):
                merged_tag = "Yes" if chunk.is_merged else "No"
                lines.append("| {} | {} | {:.1f} | {} | {} |".format(
                    i + 1, chunk.source_section, chunk.size_kb,
                    chunk.line_count, merged_tag
                ))
            lines.append("")

        # Operations
        if context.operations:
            lines.append("## Operations")
            lines.append("")
            lines.append("| Type | Target | Detail |")
            lines.append("|------|--------|--------|")
            for op in context.operations:
                lines.append("| {} | {} | {} |".format(
                    op.operation, op.target, op.detail
                ))
            lines.append("")

        # Validation results
        if context.validation_results:
            lines.append("## Validation")
            lines.append("")
            all_pass = all(context.validation_results.values())
            for check, result in context.validation_results.items():
                status = "pass" if result else "FAIL"
                lines.append("- **{}**: {}".format(check, status))
            lines.append("")
            if all_pass:
                lines.append("All checks passed.")
            else:
                lines.append("Some checks failed.")
            lines.append("")

        # File list
        if context.chunks:
            lines.append("## Output Files")
            lines.append("")
            for i, chunk in enumerate(context.chunks):
                from doc_segmenter.utils import sanitize_filename
                filename = "chunk-{:02d}-{}.md".format(
                    i + 1, sanitize_filename(chunk.source_section)
                )
                lines.append("{}. {} ({:.1f} KB)".format(
                    i + 1, filename, chunk.size_kb
                ))
            lines.append("")

        report_path = os.path.join(output_dir, "report.md")
        with open(report_path, "w", encoding="utf-8") as f:
            f.write("\n".join(lines))

File: yiyue31-translator/scripts/doc_segmenter/src/reporter.ts
/**
 * ReportGenerator implementation.
 *
 * Generates report.md from SplitContext with source info, split results,
 * operations, validation results, and file list.
 */

import { mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import { type SplitContext } from "./models";
import { sanitizeFilename } from "./utils";

export class ReportGeneratorImpl {
  /**
   * Generate report.md from the full pipeline context.
   *
   * @param context - The complete SplitContext with all pipeline data.
   */
  generateReport(context: SplitContext): void {
    const outputDir = context.outputDir;
    mkdirSync(outputDir, { recursive: true });

    const lines: string[] = [];
    lines.push("# File Split Report");
    lines.push("");

    // Source file information
    lines.push("## Source Information");
    lines.push("");
    lines.push("| Property | Value |");
    lines.push("|----------|-------|");
    lines.push(`| File | ${context.sourceInfo.filePath} |`);
    lines.push(`| Size | ${context.sourceInfo.fileSize.toFixed(1)} KB |`);
    lines.push(`| Lines | ${context.sourceInfo.fileLines} |`);
    lines.push(`| Chars | ${context.sourceInfo.fileChars} |`);
    lines.push(`| Encoding | ${context.sourceInfo.fileEncoding} |`);
    lines.push("");

    // Split results
    lines.push("## Split Results");
    lines.push("");
    lines.push(`Total chunks: ${context.chunks.length}`);
    lines.push("");

    if (context.chunks.length > 0) {
      lines.push("| # | Section | Size (KB) | Lines | Merged |");
      lines.push("|---|---------|-----------|-------|--------|");
      for (let i = 0; i < context.chunks.length; i++) {
        const chunk = context.chunks[i];
        const mergedTag = chunk.isMerged ? "Yes" : "No";
        lines.push(`| ${i + 1} | ${chunk.sourceSection} | ${chunk.sizeKb.toFixed(1)} | ${chunk.lineCount} | ${mergedTag} |`);
      }
      lines.push("");
    }

    // Operations
    if (context.operations.length > 0) {
      lines.push("## Operations");
      lines.push("");
      lines.push("| Type | Target | Detail |");
      lines.push("|------|--------|--------|");
      for (const op of context.operations) {
        lines.push(`| ${op.operation} | ${op.target} | ${op.detail} |`);
      }
      lines.push("");
    }

    // Validation results
    const validationKeys = Object.keys(context.validationResults);
    if (validationKeys.length > 0) {
      lines.push("## Validation");
      lines.push("");
      const allPass = validationKeys.every(k => context.validationResults[k]);
      for (const check of validationKeys) {
        const result = context.validationResults[check];
        const status = result ? "pass" : "FAIL";
        lines.push(`- **${check}**: ${status}`);
      }
      lines.push("");
      if (allPass) {
        lines.push("All checks passed.");
      } else {
        lines.push("Some checks failed.");
      }
      lines.push("");
    }

    // File list
    if (context.chunks.length > 0) {
      lines.push("## Output Files");
      lines.push("");
      for (let i = 0; i < context.chunks.length; i++) {
        const chunk = context.chunks[i];
        const filename = `chunk-${String(i + 1).padStart(2, "0")}-${sanitizeFilename(chunk.sourceSection)}.md`;
        lines.push(`${i + 1}. ${filename} (${chunk.sizeKb.toFixed(1)} KB)`);
      }
      lines.push("");
    }

    const reportPath = join(outputDir, "report.md");
    writeFileSync(reportPath, lines.join("\n"), "utf-8");
  }
}
